Remove provider guards, change RabbitWorkState to use IMessage, update versions of RabbitMQ projects

tiggerite opened the pull request:
The compression and encryption providers aren’t mandatory for these steps where they are guarded against null. RabbitWorkState should probably be SendMessage/SendMessageSent with IMessage type now. Also updated all the project versions so nothing is out of date (hopefully) and ready to NuGet publish

tiggerite committed:
Remove guards around providers

These aren’t mandatory for the steps they are attached to

tiggerite committed:
Change SendLetter properties to SendMessage and IMessage type, update versions of RabbitMQ csprojs

Commit:
Adding a missing feature between Topologer and Consumer auto-topology creation.

tristan-hyams commented on src/HouseofCat.RabbitMQ.Dataflows/ConsumerDataflow.cs:
```diff
@@ -173,7 +173,6 @@ public ConsumerDataflow<TState> WithReadyToProcessBuffer(int boundedCapacity)
             int? boundedCapacity = null)
         {
             Guard.AgainstNull(suppliedStep, nameof(suppliedStep));
-            Guard.AgainstNull(_encryptionProvider, nameof(_encryptionProvider));
```
Good fix.

houseofcat commented:
Added prod support fix for a bug that landed on my desk.

Utf8Json passed current requirements for File deserialization. Dictionary<string, object>() is now supported for Args both in Topologer and Consumer Topology Auto Creation.

Commit:
Version bump. Number chosen to avoid assembly version fat finger mistake.

Commit:
Breaking change, making MessageId a first class citizen like it is in RabbitMQ Client/Server.

Commit:
Update HouseofCat.RabbitMQ.csproj

Commit:
Dependency updates.

houseofcat commented:
I avoided interfaces for properties a longtime ago due to various issues with deserialization.

Currently working on solves via Integration Tests which needed updates.
`NotSupportedException: Deserialization of interface types is not supported. Type 'HouseofCat.RabbitMQ.IMetadata'.`

Commit:
Fixes.

tiggerite reviewed (commented):
I actually changed the property originally to MessageId in my first PR but went safe with the function, this is much cleaner.

tiggerite commented on src/HouseofCat.RabbitMQ/Messages/Letter.cs:
```diff
@@ -95,7 +95,7 @@ public Letter(string exchange, string routingKey, byte[] data, string id, byte p
         {
             var clone = this.Clone<Letter>();
-            clone.LetterMetadata = LetterMetadata.Clone<LetterMetadata>();
+            clone.LetterMetadata = (LetterMetadata)LetterMetadata.Clone<LetterMetadata>();
```
We could change the extension method too for cleaner code:

`public static T Clone<T>(this IMetadata metadata) where T: IMetadata, new() `

tiggerite replied:
Oh the markup removed the T generic around the Clone, but you get the gist hopefully..

tristan-hyams replied:
Good point, keeping with the current trend.

Commit:
Fix.

houseofcat commented:
Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.

Updating RabbitMQ server / Erlang locally since they are out of date.

Commit:
Update PublisherTests.cs

tiggerite commented:
> Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> 
> Updating RabbitMQ server / Erlang locally since they are out of date.



> Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> 
> Updating RabbitMQ server / Erlang locally since they are out of date.

Hmm none of them rely on Clone of the metadata do they? Because with that being new T now its probably pretty slow, on a par with Activator.CreateInstance. Could be swapped out for your Reflection library New version actually as it’s a default constructor.

houseofcat commented:
> > Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> > Updating RabbitMQ server / Erlang locally since they are out of date.
> 
> > Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> > Updating RabbitMQ server / Erlang locally since they are out of date.
> 
> Hmm none of them rely on Clone of the metadata do they? Because with that being new T now its probably pretty slow, on a par with Activator.CreateInstance. Could be swapped out for your Reflection library New version actually as it’s a default constructor.

Already did :)

tiggerite commented on src/HouseofCat.RabbitMQ/Messages/PublishReceipt.cs:
```diff
@@ -3,18 +3,17 @@ namespace HouseofCat.RabbitMQ
     public interface IPublishReceipt
     {
         bool IsError { get; set; }
+        public string MessageId { get; set; }
```
Oops the public probably shouldn’t be on an interface right?

tristan-hyams replied:
No, its a mistake I will fix.

Publishing bug had me frazzled.

Commit:
Test fixes.

Important Pub/Sub test failed due to XUnit2 Test ThreadPool MaxConcurrency getting tripped from the library and unable to resolve itself as the main Test invocation thread was stuck in While ( await Task.Delay(1) ) ;

Adding Task.Yield() in the test method before the loops fixed test and the lost publishes. Prior, the test never finished and would time out. Debugging the behavior left me confused because it never finished publishing which has never happened before.

Suffice to say, be careful with high performant integration tests... they are not benchmarks don't treat them that way.

Commit:
Update PublishReceipt.cs

houseofcat commented:
Important Pub/Sub test failed due to XUnit2 Test ThreadPool MaxConcurrency getting tripped from the library and unable to resolve itself as the main Test invocation thread was stuck in While ( await Task.Delay(1) ) ;

Adding Task.Yield() in the test method before the loops fixed test and the lost publishes. Prior, the test never finished and would time out. Debugging the behavior left me confused because it never finished publishing which has never happened before.

Suffice to say, be careful with high performant integration tests... they are not benchmarks don't treat them that way.

tiggerite commented:
> Adding Task.Yield() in the test method before the loops fixed test and the lost publishes.

That’s a relief, I thought HouseofCat RabbitMQ Client was going to make a comeback ;)
